edb/server/protocol/auth_ext/oauth.py

```diff
@@ -34,7 +35,8 @@ def __init__(
         if edgedb_test_url:
             self.edgedb_orig_base_url = urllib.parse.quote(base_url, safe='')
             base_url = edgedb_test_url
-        super().__init__(*args, base_url=base_url, **kwargs)
+        cache = httpx_cache.AsyncCacheControlTransport()
```

Comment:
If I understand your code correctly, before we share client instances among requests, this cache is only available within the handling of one user request, right? Like we may call get_oidc_config() multiple times but make an actual HTTP call only once.

This is fine now, but when we want to cache the OIDC config for a longer time (probably by sharing the same client instance), I'm not sure if `httpx_cache` is sufficient - HTTP cache is rather complicated, according to my likely outdated knowledge. For example, I don't see the use of "ETag" or "HEAD request" in `httpx_cache`.

Reply:
> If I understand your code correctly, before we share client instances among requests, this cache is only available within the handling of one user request, right? Like we may call get_oidc_config() multiple times but make an actual HTTP call only once.

I honestly don't know, I'm glad you asked! I was hoping the cache was shared by every client that uses the `http_factory`, but if that doesn't seem to be the case here, we're not getting the benefit I was hoping for with the cache. The `http_factory` related stuff was introduced by @1st1 to help with testing external HTTP requests, so maybe Yury can shed some light here?

> I'm not sure if httpx_cache is sufficient - HTTP cache is rather complicated, according to my likely outdated knowledge. For example, I don't see the use of "ETag" or "HEAD request" in httpx_cache.

Yeah, I was also surprised that `httpx_cache` didn't seem to leverage `etag`, but it _does_ respect the cache-control headers and correctly handle caching based on HTTP verb and status code which is better than a more naive LRU cache. Worst case scenario, it downloads duplicate data more than it might if there was a working etag system, so doesn't seem like a hard blocker.